panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,513 @@
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn PantherImuSensor::on_cleanup(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Cleaning up Panther Imu");
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn PantherImuSensor::on_activate(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Activating Panther Imu");
+  rclcpp::NodeOptions ros_interface_options;
+  panther_imu_ros_interface_ = std::make_unique<PantherImuRosInterface>(
+    std::bind(&PantherImuSensor::Calibrate, this),
+    "panther_imu_"
+    "node");
+
+  RCLCPP_INFO_STREAM(
+    logger_, "Connecting to Phidgets Spatial serial " << params_.serial << ", hub port "
+                                                      << params_.hub_port << " ...");
+
+  try {
+    spatial_ = std::make_unique<phidgets::Spatial>(
+      params_.serial, params_.hub_port, false,
+      std::bind(
+        &PantherImuSensor::SpatialDataCallback, this, std::placeholders::_1, std::placeholders::_2,
+        std::placeholders::_3, std::placeholders::_4),
+      nullptr, std::bind(&PantherImuSensor::SpatialAttachCallback, this),
+      std::bind(&PantherImuSensor::SpatialDetachCallback, this));
+
+    RCLCPP_INFO_STREAM(logger_, "Connected to serial " << spatial_->getSerialNumber());
+    imu_connected_ = true;
+
+    spatial_->setDataInterval(params_.data_interval_ms);
+
+    Calibrate();
+
+    ConfigureCompassParams();
+    ConfigureHeating();
+  } catch (const phidgets::Phidget22Error & err) {
+    RCLCPP_ERROR_STREAM(logger_, "Spatial: " << err.what());
+    return CallbackReturn::ERROR;
+  }
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn PantherImuSensor::on_deactivate(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Deactivating Panther Imu");
+  panther_imu_ros_interface_.reset();
+  spatial_.reset();
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn PantherImuSensor::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Shutting down Panther Imu");
+  panther_imu_ros_interface_.reset();
+  spatial_.reset();
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn PantherImuSensor::on_error(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Handling Panther Imu error");
+  panther_imu_ros_interface_.reset();
+  spatial_.reset();
+
+  return CallbackReturn::SUCCESS;
+}
+
+std::vector<StateInterface> PantherImuSensor::export_state_interfaces()
+{
+  std::vector<StateInterface> state_interfaces;
+  RCLCPP_INFO_STREAM(
+    logger_, "State interfaces count: " << info_.sensors[0].state_interfaces.size());
+
+  for (auto i = 0u; i < info_.sensors[0].state_interfaces.size(); i++) {
+    RCLCPP_INFO_STREAM(
+      logger_, "State interface name: " << info_.sensors[0].state_interfaces[i].name);
+
+    state_interfaces.emplace_back(StateInterface(
+      info_.sensors[0].name, info_.sensors[0].state_interfaces[i].name, &imu_sensor_state_[i]));
+  }
```

Comment:
```suggestion
  for (size_t i = 0; i < info_.sensors[0].state_interfaces.size(); i++) {
    state_interfaces.emplace_back(
      info_.sensors[0].name, info_.sensors[0].state_interfaces[i].name, &imu_sensor_state_[i]);
  }
```
Could we also use kImuInterfacesSize instead of `kImuInterfacesSize`

Reply:
I don't get it. Did you mean to use `kImuInterfacesSize` instead of `info_.sensors[0].state_interfaces.size()`?

Reply:
Exactly, sorry xd

Reply:
It can be. I chech if the size is correct in `CheckInterfaces` function